### PR TITLE
Addition of HiLU (Hinge Linear Unit ) parametric nonlinearity

### PR DESCRIPTION
**Pull request by kadeng**
A flexible and simple parametric invertitble nonlinearity, which separately parameterizes the slope for negative and positive elements. Efficient calculation of forward/inverse and logabsdet due to parametrization in log-space.

### CODE REVIEW

File: nflows/transforms/nonlinearities.py [hunk 1]
```diff
@@ -447,3 +447,53 @@ def forward(self, inputs, context=None):
 
     def inverse(self, inputs, context=None):
         return self._spline(inputs, inverse=True)
+
+class HiLU(Transform):
+
+    def __init__(self, param_shape, init_log_range=0.0):
```
Comment (arturbekasov) [inline nflows/transforms/nonlinearities.py:453]: Other nonlinearities call this parameter `shape`. Might not be the best name, but I propose sticking to that for consistency.

File: nflows/transforms/nonlinearities.py [hunk 2]
```diff
@@ -447,3 +447,53 @@ def forward(self, inputs, context=None):
...
+        """
+        Hinge Linear Unit
+
+        similar to a PReLU, except that the slope is parameterized separately
+        for both positive and negative inputs/outputs
+        :param param_shape: Tuple of integers - shape of the parameter tensors including batch dim,
+                            must be broadcastable to input/output shape
+        :param init_range: Optional float: Randomize initial (log-scale) parameters by drawing them from a random
```
Comment (arturbekasov) [inline nflows/transforms/nonlinearities.py:461]: `init_log_range`?

File: nflows/transforms/nonlinearities.py [hunk 3]
```diff
@@ -447,3 +447,53 @@ def forward(self, inputs, context=None):
...
+                           uniform in the interval [-init_range, +init_range]
+        """
+        super().__init__()
+        self.param_shape = param_shape
+        self.init_log_range = init_log_range
+        self.pscale = nn.Parameter(torch.zeros(size=param_shape))
+        self.nscale = nn.Parameter(torch.zeros(size=param_shape))
+        if init_log_range>0.0:
+            with torch.no_grad():
+                self.pscale.data.copy_(torch.rand_like(self.pscale) * 2.0 * init_log_range - init_log_range)
```
Comment (arturbekasov) [inline nflows/transforms/nonlinearities.py:471]: `torch.nn.init.uniform_` is a nicer way of doing this (and it takes care of doing `no_grad`).

File: nflows/transforms/nonlinearities.py [hunk 4]
```diff
@@ -447,3 +447,53 @@ def forward(self, inputs, context=None):
...
+    def forward(self, input, context=None):
+        pscale = self.pscale.exp()
+        nscale = self.nscale.exp()
+        output = torch.where(input.ge(0.0), input * pscale, input * nscale)
+        log_grad = torch.where(input.ge(0.0), self.pscale, self.nscale) # gradient of natural exp function is itself,
+                                                                        # log of that are our parameters
+        jacobian_log_abs_det = torch.sum(log_grad.view((log_grad.shape[0],-1)), dim=1)
+        return output, jacobian_log_abs_det
+
+    def _forward(self, input):
```
Comment (arturbekasov) [inline nflows/transforms/nonlinearities.py:491]: Don't see where `_forward` and `_inverse` functions are used?

File: nflows/transforms/nonlinearities.py (hunk 1 above)
Comment (kadeng) [reply nflows/transforms/nonlinearities.py:453]: Makes sense.

File: nflows/transforms/nonlinearities.py (hunk 4 above)
Comment (kadeng) [reply nflows/transforms/nonlinearities.py:491]: They are a leftover from how my library works, I don't always need the jacobian log determinant. They can be safely removed.

**Comment by kadeng** [comment]
You are right that there is no paper about it. It's a pretty trivial nonlinearity which I came up with. Here is why I consider it useful:

1.) it can be used instead of a diagonal matrix transformation, sandwiched between two orthogonal transforms. That way, you can initialize those based on an SVD decomposition of any linear matrix transform (including the identity matrix or something very close to that) 
2. So you can start training a very deep network as an effectively shallow linear architecture and introduce nonlinearities gradually via gradient descent. 
3.) It is numerically pretty stable in both directions as long as the parameters don't diverge too much from 0.
4.) It is computationally neat and tidy, and has a nice parametrization. 

On a side note, I am on holidays by now, with limited access to a computer for the next three weeks. So feel free to adapt the code.

**Comment by arturbekasov** [comment]
Closing old PRs: @kadeng please let me know if you're still interested in merging this in and I can re-open.
